File: Chagok/frontend/src/types/evidence.ts
export type EvidenceType = 'text' | 'image' | 'audio' | 'video' | 'pdf';
export type EvidenceStatus = 'uploading' | 'queued' | 'processing' | 'review_needed' | 'completed' | 'failed';
export type SpeakerType = '원고' | '피고' | '제3자' | 'unknown';

/**
 * 민법 840조 이혼 사유 카테고리
 * Korean Civil Code Article 840 - Grounds for Divorce
 */
export type Article840Category =
    | 'adultery'                  // 제1호: 배우자의 부정행위
    | 'desertion'                 // 제2호: 악의의 유기
    | 'mistreatment_by_inlaws'    // 제3호: 배우자 직계존속의 부당대우
    | 'harm_to_own_parents'       // 제4호: 자기 직계존속 피해
    | 'unknown_whereabouts'       // 제5호: 생사불명 3년
    | 'irreconcilable_differences' // 제6호: 혼인 지속 곤란사유
    | 'general';                  // 일반 증거

/**
 * Article 840 태깅 결과
 */
export interface Article840Tags {
    categories: Article840Category[];
    confidence: number; // 0.0 ~ 1.0
    matchedKeywords: string[];
}

/**
 * AI 분석 인사이트
 */
export interface AIInsight {
    id: string;
    evidenceId: string;
    type: 'summary' | 'legal_relevance' | 'risk_factor' | 'recommendation';
    title: string;
    content: string;
    confidence: number;
    createdAt: string;
}

/**
 * 015-evidence-speaker-mapping: 개별 화자 매핑 항목
 */
export interface SpeakerMappingItem {
    party_id: string;
    party_name: string;
}

/**
 * 015-evidence-speaker-mapping: 화자 매핑 객체 (화자명 → 인물 정보)
 */
export type SpeakerMapping = Record<string, SpeakerMappingItem>;

/**
 * 015-evidence-speaker-mapping: 화자 매핑 업데이트 요청
 */
export interface SpeakerMappingUpdateRequest {
    speaker_mapping: SpeakerMapping;
}

/**
 * 015-evidence-speaker-mapping: 화자 매핑 업데이트 응답
 */
export interface SpeakerMappingUpdateResponse {
    evidence_id: string;
    speaker_mapping: SpeakerMapping | null;
    updated_at: string | null;
    updated_by: string | null;
}

export interface Evidence {
    id: string;
    caseId: string;
    filename: string;
    type: EvidenceType;
    status: EvidenceStatus;
    uploadDate: string; // ISO Date string
    summary?: string;
    size: number;
    downloadUrl?: string;
    content?: string; // STT/OCR 원문 텍스트

    // AI Worker generated fields (optional - populated after processing)
    speaker?: SpeakerType;
    labels?: string[]; // AI-generated labels (e.g., '폭언', '불륜', '유책사유')
    timestamp?: string; // ISO Date string - when the evidence was created/recorded
    s3Key?: string; // S3 storage path
    qdrantId?: string; // Vector store ID for RAG

    // Article 840 analysis (optional - populated after AI processing)
    article840Tags?: Article840Tags;
    insights?: AIInsight[];

    // 014-ui-settings-completion: Legal evidence numbering
    legalNumber?: string;  // 갑제1호증, 을제2호증, etc.
    submittedBy?: 'plaintiff' | 'defendant' | 'third_party';  // 갑/을/병
    evidenceTimestamp?: string;  // ISO Date - when evidence was created (not uploaded)

    // 015-evidence-speaker-mapping: 화자 매핑 필드
    speakerMapping?: SpeakerMapping;
    speakerMappingUpdatedAt?: string;
    speakerMappingUpdatedBy?: string;
    hasSpeakerMapping?: boolean; // 목록에서 매핑 상태 표시용
}

File: Chagok/backend/app/db/schemas/evidence.py
"""
Evidence Schemas - Upload, Analysis, Review
"""

from pydantic import BaseModel, Field
from typing import Optional, Dict
from datetime import datetime
from enum import Enum


# ============================================
# Article 840 Categories
# ============================================
class Article840Category(str, Enum):
    """
    민법 840조 이혼 사유 카테고리

    Korean Civil Code Article 840 - Grounds for Divorce
    """
    ADULTERY = "adultery"  # 제1호: 배우자의 부정행위
    DESERTION = "desertion"  # 제2호: 악의의 유기
    MISTREATMENT_BY_INLAWS = "mistreatment_by_inlaws"  # 제3호: 배우자 직계존속의 부당대우
    HARM_TO_OWN_PARENTS = "harm_to_own_parents"  # 제4호: 자기 직계존속 피해
    UNKNOWN_WHEREABOUTS = "unknown_whereabouts"  # 제5호: 생사불명 3년
    IRRECONCILABLE_DIFFERENCES = "irreconcilable_differences"  # 제6호: 혼인 지속 곤란사유
    GENERAL = "general"  # 일반 증거 (특정 조항에 해당하지 않음)


class Article840Tags(BaseModel):
    """Article 840 tagging result schema"""
    categories: list[Article840Category] = Field(default_factory=list)
    confidence: float = Field(default=0.0, ge=0.0, le=1.0)
    matched_keywords: list[str] = Field(default_factory=list)


# ============================================
# Upload Schemas
# ============================================
class PresignedUrlRequest(BaseModel):
    """Presigned URL request schema"""
    case_id: str
    filename: str
    content_type: str


class PresignedUrlResponse(BaseModel):
    """Presigned URL response schema"""
    upload_url: str
    fields: dict
    evidence_temp_id: str
    s3_key: str


class ExifMetadataInput(BaseModel):
    """EXIF metadata input from client-side extraction"""
    gps_latitude: Optional[float] = Field(None, description="GPS latitude in decimal degrees")
    gps_longitude: Optional[float] = Field(None, description="GPS longitude in decimal degrees")
    gps_altitude: Optional[float] = Field(None, description="GPS altitude in meters")
    datetime_original: Optional[str] = Field(None, description="Original capture datetime (ISO format)")
    camera_make: Optional[str] = Field(None, description="Camera manufacturer")
    camera_model: Optional[str] = Field(None, description="Camera model")


class UploadCompleteRequest(BaseModel):
    """Upload complete request schema"""
    case_id: str
    evidence_temp_id: str
    s3_key: str
    file_size: int = 0  # File size in bytes
    note: Optional[str] = None
    exif_metadata: Optional[ExifMetadataInput] = Field(
        None,
        description="EXIF metadata extracted from image on client side (for detective uploads)"
    )


class UploadCompleteResponse(BaseModel):
    """Upload complete response schema"""
    evidence_id: str
    case_id: str
    filename: str
    s3_key: str
    status: str  # pending (waiting for AI processing)
    review_status: Optional[str] = None  # pending_review for client uploads, None for internal uploads
    created_at: datetime


# ============================================
# Evidence List/Detail Schemas
# ============================================
class EvidenceSummary(BaseModel):
    """Evidence summary schema (for list view)"""
    id: str
    case_id: str
    type: str  # text, image, audio, video, pdf
    filename: str
    size: int  # File size in bytes
    created_at: datetime
    status: str  # pending, processing, done, error
    ai_summary: Optional[str] = None  # AI-generated summary
    article_840_tags: Optional[Article840Tags] = None  # Article 840 tagging
    # 015-evidence-speaker-mapping: Speaker mapping status
    has_speaker_mapping: bool = Field(default=False, description="Whether speaker mapping is configured")


class EvidenceListResponse(BaseModel):
    """Evidence list response wrapper (matches frontend expectation)"""
    evidence: list[EvidenceSummary]
    total: int


class EvidenceReviewRequest(BaseModel):
    """Evidence review request schema (for lawyer approval)"""
    action: str = Field(..., pattern="^(approve|reject)$", description="Review action: approve or reject")
    comment: Optional[str] = Field(None, max_length=500, description="Optional review comment")


class EvidenceReviewResponse(BaseModel):
    """Evidence review response schema"""
    evidence_id: str
    case_id: str
    review_status: str  # approved, rejected
    reviewed_by: str
    reviewed_at: datetime
    comment: Optional[str] = None


class EvidenceDetail(BaseModel):
    """Evidence detail schema (for detail view with AI analysis)"""
    id: str
    case_id: str
    type: str
    filename: str
    size: int  # File size in bytes
    s3_key: str
    content_type: str
    created_at: datetime
    status: str

    # AI analysis results (available when status="done")
    ai_summary: Optional[str] = None
    labels: Optional[list[str]] = None  # Mapped from article_840_tags.categories
    insights: Optional[list[str]] = None
    content: Optional[str] = None  # Full STT/OCR text
    speaker: Optional[str] = None  # For audio/video
    timestamp: Optional[datetime] = None  # Event timestamp in evidence
    qdrant_id: Optional[str] = None  # RAG index reference (Qdrant point ID)
    article_840_tags: Optional[Article840Tags] = None  # Article 840 tagging

    # 015-evidence-speaker-mapping: Speaker mapping fields
    speaker_mapping: Optional[Dict[str, "SpeakerMappingItem"]] = None
    speaker_mapping_updated_at: Optional[datetime] = None


# ============================================
# Speaker Mapping Schemas (015-evidence-speaker-mapping)
# ============================================
class SpeakerMappingItem(BaseModel):
    """Individual speaker mapping item"""
    party_id: str = Field(..., description="PartyNode ID from relationship graph")
    party_name: str = Field(..., description="Party name for display")


class SpeakerMappingUpdateRequest(BaseModel):
    """Speaker mapping update request"""
    speaker_mapping: Dict[str, SpeakerMappingItem] = Field(
        default_factory=dict,
        description="Speaker label to party mapping. Empty dict clears mapping.",
        json_schema_extra={
            "examples": [{
                "나": {"party_id": "party_001", "party_name": "김동우"},
                "상대방": {"party_id": "party_002", "party_name": "김도연"}
            }]
        }
    )


class SpeakerMappingResponse(BaseModel):
    """Speaker mapping update response"""
    evidence_id: str
    speaker_mapping: Optional[Dict[str, SpeakerMappingItem]] = None
    updated_at: Optional[datetime] = None
    updated_by: Optional[str] = None


# Update forward reference for EvidenceDetail
EvidenceDetail.model_rebuild()
